tokenawarebandit.go

```diff
@@ -0,0 +1,614 @@
+		// We can't call getKeyspaceMetadata while holding the lock since getKeyspaceMetadata can do I/O.
+		metadata, err := t.getKeyspaceMetadata(keyspaceName)
+		if err != nil {
+			t.logger.Println(err)
+			continue
+		}
+
+		strategy := getStrategy(metadata)
+		var replicas tokenRingReplicas
+
+		if strategy != nil {
+			t.mu.RLock()
+			tokenRing := t.tokenRing
+			t.mu.RUnlock()
+			replicas = strategy.replicaMap(tokenRing)
+		}
+
+		t.mu.Lock()
+		ks.replicas = replicas
+		t.mu.Unlock()
+	}
+	if t.afterReplicasUpdated != nil {
+		t.afterReplicasUpdated(keyspaceName)
+	}
+}
+
+// SetPartitioner is called to set the partitioner used.
+func (t *TokenAwareBanditHostPolicy) SetPartitioner(partitioner string) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if t.partitioner == partitioner {
+		return
+	}
+
+	t.partitioner = partitioner
+
+	err := t.resetRing()
+	if err != nil {
+		Logger.Println(err)
+	}
+}
+
+// resetRing replaces the token ring with a new one.
+// must be called with t.mu lock held for writing.
+func (t *TokenAwareBanditHostPolicy) resetRing() error {
+	newTR, err := newTokenRing(t.partitioner, t.hostInfoSlice())
+	if err != nil {
+		return err
+	}
+	t.tokenRing = newTR
+
+	for keyspaceName := range t.keyspaces {
+		t.updateReplicas(keyspaceName)
+	}
+
+	return nil
+}
+
+// AddHost adds a single host to the policy after it was discovered/added to the cluster.
+func (t *TokenAwareBanditHostPolicy) AddHost(host *HostInfo) {
+	t.AddHosts([]*HostInfo{host})
+}
+
+// AddHosts adds multiple hosts to the policy after they were discovered/added to the cluster.
+func (t *TokenAwareBanditHostPolicy) AddHosts(hosts []*HostInfo) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	added := false
+
+	for _, host := range hosts {
+		hostID := host.HostID()
+		if hostID == "" {
+			continue
+		}
+		_, ok := t.hosts[hostID]
+		if !ok {
+			t.hosts[hostID] = &tokenAwareBanditHost{
+				hostInfo: host,
+			}
+			added = true
+		}
+	}
+
+	if !added {
+		return
+	}
+
+	err := t.resetRing()
+	if err != nil {
+		t.logger.Println(err)
+	}
+}
+
+// RemoveHost from the policy after it was removed from the cluster.
+func (t *TokenAwareBanditHostPolicy) RemoveHost(host *HostInfo) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	hostID := host.HostID()
+	if hostID == "" {
+		return
+	}
+
+	if _, ok := t.hosts[hostID]; !ok {
+		return
+	}
+
+	delete(t.hosts, hostID)
+	err := t.resetRing()
+	if err != nil {
+		t.logger.Println(err)
+	}
+}
+
+func (t *TokenAwareBanditHostPolicy) HostUp(host *HostInfo) {}
+
+func (t *TokenAwareBanditHostPolicy) HostDown(host *HostInfo) {}
+
+// findReplicas gets replicas for the query.
+// If for any reason we can't find replicas, the returned slice's length will be zero.
+// t.mu must be held at least read-only.
+func (t *TokenAwareBanditHostPolicy) findReplicas(qry ExecutableQuery) (token, []*HostInfo) {
+	if qry == nil {
+		return nil, nil
+	}
+
+	routingKey, err := qry.GetRoutingKey()
+	if err != nil || routingKey == nil {
+		return nil, nil
```

Comment:
Is it ok that we just ignore the `err` returned from `qry.GetRoutingKey()` here? Shouldn't we at least log it or something?

Reply:
Good question! Returning nil means that we'll send the query to random node. GetRoutingKey() will return nil for example for DDL statements (CREATE TABLE, etc.). I don't know how often an error could be returned here, I guess we could try to log it and remove the logging in case it will be too verbose.

Reply:
Thanks for explanation! Lets try and we shall see 👍

Reply:
Added the logging statement.